README.md
# Color Transfer via Sliced Optimal Transport

website : [https://dcoeurjo.github.io/OTColorTransfer/](https://dcoeurjo.github.io/OTColorTransfer/)

@[David Coeurjolly](https://perso.liris.cnrs.fr/david.coeurjolly/), CNRS, Université de Lyon, LIRIS


docs-source/Changelog.md
* Oct 16, 2020: Fix in openMP for macOS, new "factor" option
* Feb 13, 2020: minor fix for better OpenMP support on macos
* Nov 21, 2019: adding python code by [@iperov](https://github.com/iperov)
* Aug 30, 2019: first version (CPU balanced and partial Sliced OT)


ndTransfer.cpp
/*
 Copyright (c) 2019 CNRS
 David Coeurjolly <david.coeurjolly@liris.cnrs.fr>
 
 All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:
 
 1. Redistributions of source code must retain the above copyright notice, this
 list of conditions and the following disclaimer.
 
 2. Redistributions in binary form must reproduce the above copyright notice,
 this list of conditions and the following disclaimer in the documentation
 and/or other materials provided with the distribution.
 
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIEDi
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
 ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#include <iostream>
#include <string>
#include <random>
#include <vector>
#include <chrono>
#include <ctime>
#include <thread>
#include <sstream>
#include <assert.h>
//Command-line parsing
#include "CLI11.hpp"

//Image filtering and I/O
#define cimg_display 0  
#include "CImg.h"

//Global flag to silent verbose messages
bool silent;

typedef std::vector<double> Point;
typedef std::vector<Point> PointSet;

PointSet loadPointset(const std::string &filename)
{
  PointSet output;
  std::string line;
  std::ifstream ifs(filename,std::ifstream::in);
  auto dim=0;
  while( std::getline( ifs, line ) ) // read the file one line at a time
  {
    Point p;
    std::istringstream iss(line);
    std::vector<std::string> results(std::istream_iterator<std::string>{iss},
                                     std::istream_iterator<std::string>());
    assert( (dim==0) || (dim = results.size()));
    dim = results.size();
    
    for(auto s: results)
        p.push_back(std::stod(s));
    output.push_back(p);
  }
  ifs.close();
  std::cout<<"NbPoints = "<< output.size()<< " dimension = "<<output[0].size()<<std::endl;
  return output;
}



void regularize(const PointSet &sourceOrig,
                PointSet &source,
                const std::vector<unsigned int> dims,
                const double sigmaXY,
                const double sigmaV)
{
  
  unsigned int size = (int)sqrt(source.size());
  
  for(auto i=0 ; i < dims.size(); ++i)
  {
    cimg_library::CImg<double> transport(size, size, 1, 1);
    for(auto j=0; j<size*size; ++j)
      transport[j] = source[j][dims[i]] - sourceOrig[j][dims[i]];;
    transport.blur_bilateral(transport, sigmaXY,sigmaV);
    
    for(auto j = 0 ; j < size*size ; ++j)
      source[j][dims[i]] = std::min(1.0, std::max(0.0, sourceOrig[j][dims[i]] + transport[j]));
    
  }
}


void dumpPointset(const std::string &filename, PointSet &pset)
{
  std::ofstream ofs(filename,std::ofstream::out);
  
  for(auto &p: pset)
  {
    for(auto &v: p)
      ofs <<v<<" ";
    ofs<<std::endl;
  }
  ofs.close();
}


/// Dot product for specified dimensions
/// @param a a point
/// @param b a second point
/// @param dims the dimensions (subspace for the dot product)
/// @return the dot product
double dot(const Point &a, const Point &dir, const std::vector<unsigned int> &dims)
{
  double res=0.0;
  for (auto i = 0; i < dims.size(); ++i)
    res+= a[ dims[i] ] * dir[ dims[i] ];
  return res;
}



void slicedTransfer(PointSet &source,
                    const PointSet &target,
                    const std::vector<unsigned int> &dims,
                    const int nbSteps,
                    const int batchSize)
{
  
  //Random generator init to draw random line directions
  std::mt19937 gen;
  gen.seed(10);
  std::normal_distribution<double> dist{0.0,1.0};
  std::uniform_real_distribution<double> unif(0.0,1.0);
  auto N = source.size();
  
  assert(source.size()==target.size());
  
  //Advection vector
  Point zero(source[0].size(), 0.0);
  PointSet advect(N, zero);
  
  //To store the 1D projections
  std::vector<double> projsource(N);
  std::vector<double> projtarget(N);
  //point Id
  std::vector<unsigned int> idSource(N);
  std::vector<unsigned int> idTarget(N);
  
  //Lambda expression for the comparison of points in RGB
  //according to their projections
  auto lambdaProjSource = [&projsource](unsigned int a, unsigned int b) {return projsource[a] < projsource[b]; };
  auto lambdaProjTarget = [&projtarget](unsigned int a, unsigned int b) {return projtarget[a] < projtarget[b]; };
  
  for(auto i=0; i <N ; ++i)
  {
    idSource[i]=i;
    idTarget[i]=i;
  }
  
  for(auto step =0 ; step < nbSteps; ++step)
  {
    for(auto batch = 0; batch < batchSize; ++batch )
    {
      //Random direction
      Point directions( source[0].size() , 0.0 );
      double norm=0.0;
     
      if (dims.size()==1)
        directions[dims[0]] = unif(gen);
      else
      {
        for(auto i = 0; i < dims.size(); ++i  )
        {
          directions[dims[i]] = dist(gen);
          norm += directions[dims[i]] * directions[dims[i]];
        }
        norm = std::sqrt(norm);
        for(auto i = 0; i < dims.size(); ++i  )
          directions[dims[i]] /= norm;
      }
     
     if (!silent)
      {
        std::cout<<"Slice "<<step<<" batch "<<batch<<"  --  ";
        for(auto i = 0; i < directions.size(); ++i  )
          std::cout<<directions[i]<<" ";
        std::cout << std::endl;
      }
      
      //We project the points
      //1D optimal transport of the projections with two sorts
      std::thread threadA([&]{for(auto i = 0; i < N; ++i)
                                 projsource[i] = dot(source[i], directions, dims);
                              std::sort(idSource.begin(), idSource.end(), lambdaProjSource); });
      
      //Parallel
      for(auto i = 0; i <N; ++i)
        projtarget[i] = dot(target[i], directions, dims);
      std::sort(idTarget.begin(), idTarget.end(), lambdaProjTarget);
      threadA.join();
      
      //We accumulate the displacements in a batch
      for(auto p = 0; p < N; ++p)
      {
        auto pix = idSource[p];
        for(auto i = 0; i < dims.size(); ++i  )
          advect[pix][dims[i]] += directions[dims[i]]  * (projtarget[idTarget[p]] - projsource[ pix ]);
      }
    }
    for(auto i = 0; i <N; ++i)
    {
      for(auto k=0; k < dims.size(); ++k)
      {
        source[i][dims[k]] += advect[i][dims[k]]/(double)batchSize;
        advect[i][dims[k]] = 0.0;
      }
    }
  }
}




int main(int argc, char **argv)
{
  CLI::App app{"colorTransfer"};
  std::string sourceImage;
  app.add_option("-s,--source", sourceImage, "Source file") ->required()->check(CLI::ExistingFile);
  std::string targetImage;
  app.add_option("-t,--target", targetImage, "Target file")->required()->check(CLI::ExistingFile);
  std::string outputImage;
  app.add_option("-o,--output", outputImage, "Output file")->required();
  unsigned int nbSteps = 3;
  app.add_option("-n,--nbsteps", nbSteps, "Number of sliced steps (3)");
  unsigned int batchSize = 1;
  app.add_option("-b,--sizeBatch", batchSize, "Number of dirtections on a batch (1)");
  bool applyRegularization = false;
  app.add_flag("-r,--regularization", applyRegularization, "Apply a regularization step of the transport plan using bilateral filter (false).");
  float sigmaXY = 16.0;
  app.add_option("--sigmaXY", sigmaXY, "Sigma parameter in the spatial domain for the bilateral regularization (16.0)");
  float sigmaV = 5.0;
  app.add_option("--sigmaV", sigmaV, "Sigma parameter in the value domain for the bilateral regularization (5.0)");
  silent = false;
  app.add_flag("--silent", silent, "No verbose messages");
 
  
  std::vector<unsigned int> dimensions;
  app.add_option("--dims", dimensions, "OT subspace");
  CLI11_PARSE(app, argc, argv);
  
 
  //Loading data
  PointSet source = loadPointset(sourceImage);
  PointSet orig = source;
  PointSet target = loadPointset(targetImage);
  
  slicedTransfer(source, target, dimensions, nbSteps, batchSize);

  if (applyRegularization)
  {
    if (!silent) std::cout<<"Applying regularization step"<<std::endl;
    regularize(orig, source, dimensions, sigmaXY, sigmaV);
  }
  //export
  dumpPointset(outputImage, source);

  exit(0);
}


colorTransferGPU.cpp
#include <iostream>
#include <string>
#include <random>
#include <vector>

#include "CLI11.hpp"

#define cimg_display 0
#include "CImg.h"
#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"

#include <boost/compute/core.hpp>
#include <boost/compute/algorithm/transform.hpp>
#include <boost/compute/container/vector.hpp>
#include <boost/compute/functional/math.hpp>
#include <boost/compute/algorithm/count_if.hpp>
#include <boost/compute/function.hpp>
#include <boost/compute/algorithm/iota.hpp>
#include <boost/compute/algorithm/stable_sort_by_key.hpp>
#include <boost/compute/algorithm/sort.hpp>
#include <boost/compute/algorithm/scatter.hpp>
#include <boost/compute/algorithm/accumulate.hpp>

#include <boost/range/adaptor/strided.hpp>
#include <boost/range/algorithm/copy.hpp>
#include <boost/assign.hpp>
#include <boost/range/algorithm.hpp>


//Global flag to silent verbose messages
bool silent;

namespace compute = boost::compute;

void slicedTransfer(std::vector<float> &source,
const std::vector<float> &target,
const int nbSteps,
const int batchSize)
{
  
  //Random generator init to draw random line directions
  std::mt19937 gen(0);
  std::normal_distribution<float> dist{0.0,1.0};
  
  const auto N = source.size() / 3;
  
  // get default OpenCL device and setup context
  compute::device device = compute::system::default_device();
  compute::context context(device);
  compute::command_queue queue(context, device);
  
  //Sending buffers
  compute::vector<float> sourceGPUx(N, context);
  compute::vector<float> sourceGPUy(N, context);
  compute::vector<float> sourceGPUz(N, context);
  compute::vector<float> targetGPUx(N, context);
  compute::vector<float> targetGPUy(N, context);
  compute::vector<float> targetGPUz(N, context);
  
  std::vector<float> tmp(N);
  for(auto i=0; i< N; ++i) tmp[i] = source[3*i] + std::abs(dist(gen))*3.1;
  compute::copy(tmp.begin(), tmp.end(), sourceGPUx.begin(), queue);
  
  for(auto i=0; i< N; ++i) tmp[i] = source[3*i+1]+  std::abs(dist(gen))*3.1;
  compute::copy(tmp.begin(), tmp.end(), sourceGPUy.begin(), queue);
  
  for(auto i=0; i< N; ++i) tmp[i] = source[3*i+2]+  std::abs(dist(gen))*3.1;
  compute::copy(tmp.begin(), tmp.end(), sourceGPUz.begin(), queue);
  
  for(auto i=0; i< N; ++i) tmp[i] = target[3*i]+  std::abs(dist(gen))*3.1;
  compute::copy(tmp.begin(), tmp.end(), targetGPUx.begin(), queue);
  
  for(auto i=0; i< N; ++i) tmp[i] = target[3*i+1]+  std::abs(dist(gen))*3.1;
  compute::copy(tmp.begin(), tmp.end(), targetGPUy.begin(), queue);
  
  for(auto i=0; i< N; ++i) tmp[i] = target[3*i+2]+  std::abs(dist(gen))*3.1;
  compute::copy(tmp.begin(), tmp.end(), targetGPUz.begin(), queue);
  
  
  //Advection vector
  compute::vector<float> advectX(N,context);
  compute::vector<float> advectY(N,context);
  compute::vector<float> advectZ(N,context);
  
  //To store the 1D projections
  compute::vector<float> projsource(N, context);
  compute::vector<float> projtarget(N, context);
  compute::vector<float> delta(N, context);
  
  //Pixel Id
  compute::vector<unsigned int> idSource(N, context);
  compute::vector<unsigned int> idTarget(N, context);
  
  
  if (!silent)
  {
    std::cout<<" TSource points: "<<std::endl;
    compute::copy(sourceGPUx.begin(), sourceGPUx.end(), std::ostream_iterator<float>(std::cout, " "), queue);
    std::cout<<std::endl;
    compute::copy(sourceGPUy.begin(), sourceGPUy.end(), std::ostream_iterator<float>(std::cout, " "), queue);
    std::cout<<std::endl;
    compute::copy(sourceGPUz.begin(), sourceGPUz.end(), std::ostream_iterator<float>(std::cout, " "), queue);
    std::cout<<std::endl;
    std::cout<<std::endl;
  }
  if (!silent)
  {
    std::cout<<" Target points: "<<std::endl;
    compute::copy(targetGPUx.begin(), targetGPUx.end(), std::ostream_iterator<float>(std::cout, " "), queue);
    std::cout<<std::endl;
    compute::copy(targetGPUy.begin(), targetGPUy.end(), std::ostream_iterator<float>(std::cout, " "), queue);
    std::cout<<std::endl;
    compute::copy(targetGPUz.begin(), targetGPUz.end(), std::ostream_iterator<float>(std::cout, " "), queue);
    std::cout<<std::endl;
    std::cout<<std::endl;
  }
  
  using boost::compute::lambda::_1;
  using boost::compute::lambda::_2;
  
  for(auto step = 0 ; step < nbSteps; ++step)
  {
    compute::fill(advectX.begin(), advectX.end(), 0.0, queue);
    compute::fill(advectY.begin(), advectY.end(), 0.0, queue);
    compute::fill(advectZ.begin(), advectZ.end(), 0.0, queue);
    for(auto batch = 0; batch < batchSize; ++batch )
    {
      //Random direction
      float dirx = dist(gen);
      float diry = dist(gen);
      float dirz = dist(gen);
      float norm = sqrt(dirx*dirx + diry*diry + dirz*dirz);
      dirx /= norm;
      diry /= norm;
      dirz /= norm;
      if (step==0)
      {
        dirx = 1.0;///= norm;
        diry = 0; //= norm;
        dirz = 0.0; ///= norm;
      }
      if (step==1)
      {
        dirx = 0.0;///= norm;
        diry = 1.0; //= norm;
        dirz = 0.0; ///= norm;
        
      }
      if (step==2)
      {
        dirx = 0.0;///= norm;
        diry = 0.0; //= norm;
        dirz = 1.0; ///= norm;
        
      }
      
      if (!silent)
        std::cout<<"Slice "<<step<<" batch "<<batch<<"  ("<<dirx<<","<<diry<<","<<dirz<<")"<<std::endl;
      
      //We project the points
      compute::transform(sourceGPUx.begin(), sourceGPUx.end(), sourceGPUy.begin(), projsource.begin(), _1*dirx+_2*diry,  queue);
      queue.finish();
      compute::transform(sourceGPUz.begin(), sourceGPUz.end(), projsource.begin(), projsource.begin(), _1*dirz + _2,  queue);
      queue.finish();
      
      // projTarget
      queue.finish();
      compute::transform(targetGPUx.begin(), targetGPUx.end(), targetGPUy.begin(), projtarget.begin(),  _1*dirx + _2*diry,  queue);
      queue.finish();
      compute::transform(targetGPUz.begin(), targetGPUz.end(), projtarget.begin(), projtarget.begin(),  _1*dirz + _2,  queue);
      queue.finish();
      
      if (!silent)
      {
        std::cout<<"Avant tri"<<std::endl;
        compute::copy(projsource.begin(), projsource.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<"Id: ";
        compute::copy(idSource.begin(), idSource.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        compute::copy(projtarget.begin(), projtarget.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<"Id: ";
        compute::copy(idTarget.begin(), idTarget.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<std::endl;
      }
      
      compute::iota(idSource.begin(), idSource.end(), 0, queue);
      queue.finish();
      compute::iota(idTarget.begin(), idTarget.end(), 0, queue);
      queue.finish();
      
      //1D optimal transport of the projections with two sorts
      compute::stable_sort_by_key(projsource.begin(),projsource.end(), idSource.begin(), queue);
      queue.finish();
      compute::stable_sort_by_key(projtarget.begin(),projtarget.end(), idTarget.begin(), queue);
      queue.finish();
      
      if (!silent)
      {std::cout<<"Apres tri"<<std::endl;
        compute::copy(projsource.begin(), projsource.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<"Id: ";
        compute::copy(idSource.begin(), idSource.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        compute::copy(projtarget.begin(), projtarget.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<"Id: ";
        compute::copy(idTarget.begin(), idTarget.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<std::endl;
        std::cout<<std::endl;
      }
      
      
      queue.finish();
      compute::transform(projtarget.begin(), projtarget.end(), projsource.begin(),  delta.begin(), compute::minus<float>(), queue);
      queue.finish();
      //delta[i] is the shift for the i-th point along the projection (idSource[i] point in Source)
      
      //Reorder delta[i]
      if (!silent)
      {
        std::cout<<"scatter"<<std::endl;
        compute::copy(delta.begin(), delta.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<"Id: ";compute::copy(idSource.begin(), idSource.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        
      }
      queue.finish();
      
      queue.finish();
      compute::scatter(delta.begin(), delta.end(), idSource.begin(), delta.begin(), queue);
      queue.finish();
      if (!silent)
      {
        std::cout<<"scatter res(source) ";
        compute::copy(delta.begin(), delta.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
      }
      
      queue.finish();
      compute::transform(advectX.begin(), advectX.end(), delta.begin(), advectX.begin(), _1 + _2*dirx, queue);
      queue.finish();
      compute::transform(advectY.begin(), advectY.end(), delta.begin(), advectY.begin(), _1 + _2*diry, queue);
      queue.finish();
      compute::transform(advectZ.begin(), advectZ.end(), delta.begin(), advectZ.begin(), _1 + _2*dirz, queue);
      queue.finish();
      
      if (!silent)
      {
        std::cout<<" Advect vectors: "<<std::endl;
        compute::copy(advectX.begin(), advectX.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        compute::copy(advectY.begin(), advectY.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        compute::copy(advectZ.begin(), advectZ.end(), std::ostream_iterator<float>(std::cout, " "), queue);
        std::cout<<std::endl;
        std::cout<<std::endl;
      }
    }
    //Advection
    compute::transform(sourceGPUx.begin(), sourceGPUx.end(),advectX.begin(),sourceGPUx.begin(), _1 + _2/(float)batchSize, queue);
    queue.finish();
    compute::transform(sourceGPUy.begin(), sourceGPUy.end(),advectY.begin(),sourceGPUy.begin(), _1 + _2/(float)batchSize, queue);
    queue.finish();
    compute::transform(sourceGPUz.begin(), sourceGPUz.end(),advectZ.begin(),sourceGPUz.begin(), _1 + _2/(float)batchSize, queue);
    queue.finish();
    
    if (!silent)
    {
      std::cout<<" Source advected points: "<<std::endl;
      compute::copy(sourceGPUx.begin(), sourceGPUx.end(), std::ostream_iterator<float>(std::cout, " "), queue);
      std::cout<<std::endl;
      compute::copy(sourceGPUy.begin(), sourceGPUy.end(), std::ostream_iterator<float>(std::cout, " "), queue);
      std::cout<<std::endl;
      compute::copy(sourceGPUz.begin(), sourceGPUz.end(), std::ostream_iterator<float>(std::cout, " "), queue);
      std::cout<<std::endl;
      std::cout<<std::endl;
    }
    
  }
  
  //Final copy to the CPU
  queue.finish();
  std::vector<float> sx(N);
  std::vector<float> sy(N);
  std::vector<float> sz(N);
  queue.finish();
  compute::copy(sourceGPUx.begin(), sourceGPUx.end(), sx.begin(), queue);
  queue.finish();
  compute::copy(sourceGPUy.begin(), sourceGPUy.end(), sy.begin(), queue);
  queue.finish();
  compute::copy(sourceGPUz.begin(), sourceGPUz.end(), sz.begin(), queue);
  for(auto i=0; i < N; ++i)
  {
    source[3*i]   = sx[i];
    source[3*i+1] = sy[i];
    source[3*i+2] = sz[i];
  }
}

int main(int argc, char **argv)
{
  CLI::App app{"colorTransfer"};
  std::string sourceImage="pexelAred.png";
  app.add_option("-s,--source", sourceImage, "Source image");
  std::string targetImage="pexelBred.png";
  app.add_option("-t,--target", targetImage, "Target image");
  std::string outputImage= "output.png";
  app.add_option("-o,--output", outputImage, "Output image");
  unsigned int nbSteps = 3;
  app.add_option("-n,--nbsteps", nbSteps, "Number of sliced steps (3)");
  unsigned int batchSize = 1;
  app.add_option("-b,--sizeBatch", batchSize, "Number of dirtections on a batch (1)");
  bool applyRegularization = false;
  app.add_flag("-r,--regularization", applyRegularization, "Apply a regularization step of the transport plan using bilateral filter (false).");
  float sigmaXY = 16.0;
  app.add_option("--sigmaXY", sigmaXY, "Sigma parameter in the spatial domain for the bilateral regularization (16.0)");
  float sigmaV = 5.0;
  app.add_option("--sigmaV", sigmaV, "Sigma parameter in the value domain for the bilateral regularization (5.0)");
  silent = false;
  app.add_flag("--silent", silent, "No verbose messages");
  CLI11_PARSE(app, argc, argv);
  
  //Image loading
  int width,height, nbChannels;
  unsigned char *source = stbi_load(sourceImage.c_str(), &width, &height, &nbChannels, 0);
  if (!silent) std::cout<< "Source image: "<<width<<"x"<<height<<"   ("<<nbChannels<<")"<< std::endl;
  int width_target,height_target, nbChannels_target;
  unsigned char *target = stbi_load(targetImage.c_str(), &width_target, &height_target, &nbChannels_target, 0);
  if (!silent) std::cout<< "Target image: "<<width_target<<"x"<<height_target<<"   ("<<nbChannels_target<<")"<< std::endl;
  
  if ((width != width_target) || (height_target != height) || (nbChannels!=nbChannels_target))
  {
    std::cout<< "Image sizes do not match. "<<std::endl;
    exit(1);
  }
  if (nbChannels <3)
  {
    std::cout<< "Input images must be color images."<<std::endl;
    exit(1);
  }
  
  std::vector<float> sourcefloat(width*height*nbChannels);
  std::vector<float> targetfloat(width*height*nbChannels);
#pragma omp parallel for
  for(auto i = 0 ; i <width*height*nbChannels; ++i)
  {
    sourcefloat[i] = static_cast<float>(source[i]);
    targetfloat[i] = static_cast<float>(target[i]);
  }
  
  //Main computation
  slicedTransfer(sourcefloat, targetfloat, nbSteps, batchSize);
  
  
  //Output
  std::vector<unsigned char> output(width*height*nbChannels);
  if (applyRegularization)
  {
    //Regularization of the transport plan (optional)
    // (bilateral filter of the difference)
    if (!silent) std::cout<<"Applying regularization step"<<std::endl;
    cimg_library::CImg<float> transport(width, height, 1, 3);
    for(auto i=0; i<width*height; ++i)
    {
      transport[i] = sourcefloat[3*i] - static_cast<float>(source[3*i]);
      transport[i+ width*height] = sourcefloat[3*i+1] - static_cast<float>(source[3*i+1]);
      transport[i+2*width*height] = sourcefloat[3*i+2] - static_cast<float>(source[3*i+2]);
    }
    transport.blur_bilateral(transport, sigmaXY,sigmaV);
    
    auto output2(output);
    for(auto i = 0 ; i < width*height ; ++i)
    {
      output[3*i]   = static_cast<unsigned char>(  std::min(255.0f, std::max(0.0f, static_cast<float>(source[3*i  ]) + transport[i])));
      output[3*i+1] = static_cast<unsigned char>(  std::min(255.0f, std::max(0.0f, static_cast<float>(source[3*i+1]) + transport[i+ width*height])));
      output[3*i+2] = static_cast<unsigned char>(  std::min(255.0f, std::max(0.0f, static_cast<float>(source[3*i+2]) + transport[i+ width*height*2])));
    }
  }
  else
  {
    for(auto i = 0 ; i < width*height*nbChannels ; ++i)
      output[i] = static_cast<unsigned char>(  std::min(255.0f, std::max(0.0f,  sourcefloat[i])));
  }
  
  //Final export
  if (!silent) std::cout<<"Exporting.."<<std::endl;
  int errcode = stbi_write_png(outputImage.c_str(), width, height, nbChannels, output.data(), nbChannels*width);
  if (!errcode)
  {
    std::cout<<"Error while exporting the resulting image."<<std::endl;
    exit(errcode);
  }
  
  stbi_image_free(source);
  stbi_image_free(target);
  exit(0);
}
